File: web/heuristics.js
// Classical Hex knowledge -- a port of alphazero_hex/heuristics.py.
// Two-distance (Anshelevich) connection strength, bridge detection, and the
// move-scoring used by the rule-based and alpha-beta agents.
'use strict';

// Node: each file is an isolated module scope, so getNei needs importing.
// Browser: all <script> tags share one global scope, so hex.js's top-level
// `const getNei` is already visible here -- this block must do nothing then,
// and in particular must not `var`-declare anything (var hoists to the
// shared global scope regardless of whether this branch runs, which would
// collide with hex.js's `const` of the same name and throw a SyntaxError).
if (typeof require !== 'undefined' && typeof getNei === 'undefined') {
  Object.assign(globalThis, require('./hex.js'));
}

const INF = 1e6;

function paddedNeighbours(n) {
  const nei = getNei(n), ncells = n * n, out = [];
  for (let i = 0; i < ncells; i++) {
    const row = new Int32Array(6).fill(ncells);
    nei[i].forEach((v, k) => { row[k] = v; });
    out.push(row);
  }
  return out;
}
const _padCache = new Map();
function getPadded(n) {
  if (!_padCache.has(n)) _padCache.set(n, paddedNeighbours(n));
  return _padCache.get(n);
}

const BRIDGE_OFFSETS = [
  [[-1, -1], [[-1, 0], [0, -1]]],
  [[1, 1], [[1, 0], [0, 1]]],
  [[-2, 1], [[-1, 0], [-1, 1]]],
  [[2, -1], [[1, 0], [1, -1]]],
  [[-1, 2], [[-1, 1], [0, 1]]],
  [[1, -2], [[1, -1], [0, -1]]],
];

function bridgeTable(n) {
  const table = [];
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      const entries = [];
      for (const [[dr, dc], carriers] of BRIDGE_OFFSETS) {
        const pr = r + dr, pc = c + dc;
        if (pr < 0 || pr >= n || pc < 0 || pc >= n) continue;
        let ok = true;
        const cells = [];
        for (const [cr, cc] of carriers) {
          const ar = r + cr, ac = c + cc;
          if (ar < 0 || ar >= n || ac < 0 || ac >= n) { ok = false; break; }
          cells.push(ar * n + ac);
        }
        if (ok) entries.push([pr * n + pc, cells[0], cells[1]]);
      }
      table.push(entries);
    }
  }
  return table;
}
const _bridgeCache = new Map();
function getBridgeTable(n) {
  if (!_bridgeCache.has(n)) _bridgeCache.set(n, bridgeTable(n));
  return _bridgeCache.get(n);
}

// Two-distance from one of `player`'s edges to every cell.
// source in {'top','bottom','left','right'}.
function twoDistance(boardFlat, n, player, source, opp) {
  const ncells = n * n, nei = getPadded(n);
  const isOwn = new Uint8Array(ncells), isOpp = new Uint8Array(ncells);
  for (let i = 0; i < ncells; i++) { isOwn[i] = boardFlat[i] === player ? 1 : 0; isOpp[i] = boardFlat[i] === opp ? 1 : 0; }

  const src = new Float64Array(ncells).fill(INF);
  for (let r = 0; r < n; r++) for (let c = 0; c < n; c++) {
    const i = r * n + c;
    if ((source === 'top' && r === 0) || (source === 'bottom' && r === n - 1) ||
        (source === 'left' && c === 0) || (source === 'right' && c === n - 1)) src[i] = 0;
  }
  for (let i = 0; i < ncells; i++) if (isOpp[i]) src[i] = INF;

  // Two candidate slots for the source edge: a board edge cannot be cut, so
  // touching it already counts as a redundant (doubly-connected) route --
  // mirrors the padding in heuristics.two_distance.
  let d = new Float64Array(ncells).fill(INF);
  const cand = new Float64Array(8);
  for (let iter = 0; iter < 2 * n + 4; iter++) {
    const nd = new Float64Array(ncells);
    let changed = false;
    for (let i = 0; i < ncells; i++) {
      if (isOpp[i]) { nd[i] = INF; if (d[i] !== INF) changed = true; continue; }
      const row = nei[i];
      for (let k = 0; k < 6; k++) cand[k] = row[k] === ncells ? INF : d[row[k]];
      cand[6] = src[i]; cand[7] = src[i];
      const arr = Array.from(cand).sort((a, b) => a - b);
      let val = isOwn[i] ? arr[0] : arr[1] + 1.0;
      if (val > d[i]) val = d[i]; // min with the previous iteration
      if (val !== d[i]) changed = true;
      nd[i] = val;
    }
    d = nd;
    if (!changed) break;
  }
  for (let i = 0; i < ncells; i++) if (d[i] > INF) d[i] = INF;
  return d;
}

function edgesFor(player, BLACK) { return player === BLACK ? ['top', 'bottom'] : ['left', 'right']; }

function connectionCost(boardFlat, n, player, BLACK, WHITE, EMPTY) {
  const opp = player === BLACK ? WHITE : BLACK;
  const [e1, e2] = edgesFor(player, BLACK);
  const d1 = twoDistance(boardFlat, n, player, e1, opp);
  const d2 = twoDistance(boardFlat, n, player, e2, opp);
  const ncells = n * n, total = new Float64Array(ncells);
  let best = INF;
  for (let i = 0; i < ncells; i++) {
    let t = d1[i] + d2[i];
    if (boardFlat[i] === EMPTY) t -= 1.0;
    if (t > INF) t = INF;
    total[i] = t;
    if (t < best) best = t;
  }
  return { potential: best, cost: total };
}

const _centralityCache = new Map();
function centrality(n) {
  if (_centralityCache.has(n)) return _centralityCache.get(n);
  const out = new Float64Array(n * n), mid = (n - 1) / 2;
  for (let r = 0; r < n; r++) for (let c = 0; c < n; c++) {
    const dist = Math.max(Math.abs(r - mid), Math.abs(c - mid)) / Math.max(mid, 1e-9);
    out[r * n + c] = 1.0 - dist;
  }
  _centralityCache.set(n, out);
  return out;
}

function moveScores(board, player, BLACK, WHITE, EMPTY, defenceWeight = 1.0) {
  const n = board.n, flat = board.board, opp = player === BLACK ? WHITE : BLACK;
  const my = connectionCost(flat, n, player, BLACK, WHITE, EMPTY);
  const op = connectionCost(flat, n, opp, BLACK, WHITE, EMPTY);
  const cen = centrality(n), ncells = n * n, score = new Float64Array(ncells);
  for (let i = 0; i < ncells; i++) {
    score[i] = -Math.min(my.cost[i], 1e5) - defenceWeight * Math.min(op.cost[i], 1e5) + 0.15 * cen[i];
    if (flat[i] !== EMPTY) score[i] = -Infinity;
  }
  return score;
}

// Static evaluation in [-1, 1] from `player`'s point of view, used as the
// leaf heuristic by the alpha-beta searcher. Mirrors heuristics.evaluate.
function evaluate(board, player, BLACK, WHITE) {
  if (board.winner) return board.winner === player ? 1.0 : -1.0;
  const n = board.n, opp = player === BLACK ? WHITE : BLACK;
  const my = connectionCost(board.board, n, player, BLACK, WHITE, 0);
  const op = connectionCost(board.board, n, opp, BLACK, WHITE, 0);
  if (my.potential >= INF) return -1.0;
  if (op.potential >= INF) return 1.0;
  const diff = op.potential - my.potential;
  return Math.tanh(diff / 3.0) * 0.99;
}

function brokenBridgeResponse(board, player, lastMove, BLACK, WHITE, EMPTY) {
  if (lastMove == null) return null;
  const n = board.n, flat = board.board, opp = player === BLACK ? WHITE : BLACK;
  if (flat[lastMove] !== opp) return null;
  const bt = getBridgeTable(n);
  for (let cell = 0; cell < n * n; cell++) {
    if (flat[cell] !== player) continue;
    for (const [partner, ca, cb] of bt[cell]) {
      if (flat[partner] !== player) continue;
      if (lastMove === ca && flat[cb] === EMPTY) return cb;
      if (lastMove === cb && flat[ca] === EMPTY) return ca;
    }
  }
  return null;
}

if (typeof module !== 'undefined') {
  module.exports = { twoDistance, connectionCost, moveScores, brokenBridgeResponse, getBridgeTable, INF, centrality, evaluate };
}

File: alphazero_hex/heuristics.py
"""Classical (non-neural) Hex knowledge: two-distance and bridges.

These power the hand-written opponents that the trained agent is measured
against.  Nothing here touches the network.

Two-distance (Anshelevich) is the standard connection-strength measure for
Hex.  Ordinary shortest path badly overrates a connection that hangs on a
single cell; two-distance charges an empty cell the *second* smallest
neighbour distance plus one, so a cell only becomes cheap when it has two
independent ways through -- exactly the redundancy that makes a Hex
connection hard to cut.
"""

from __future__ import annotations

import numpy as np

from .hex_game import BLACK, EMPTY, WHITE, HexBoard, neighbour_table

INF = 1e6

_PAD_CACHE: dict[int, np.ndarray] = {}
_BRIDGE_CACHE: dict[int, tuple] = {}

# (dr, dc) of the six bridge partners, each with the two carrier offsets that
# must stay empty for the bridge to be a real (second-order) connection.
BRIDGE_OFFSETS = (
    ((-1, -1), ((-1, 0), (0, -1))),
    ((1, 1), ((1, 0), (0, 1))),
    ((-2, 1), ((-1, 0), (-1, 1))),
    ((2, -1), ((1, 0), (1, -1))),
    ((-1, 2), ((-1, 1), (0, 1))),
    ((1, -2), ((1, -1), (0, -1))),
)


def padded_neighbours(n: int) -> np.ndarray:
    """(n*n, 6) int array; off-board entries point at a sentinel slot n*n."""
    cached = _PAD_CACHE.get(n)
    if cached is not None:
        return cached
    nei = neighbour_table(n)
    out = np.full((n * n, 6), n * n, dtype=np.int32)
    for i, nbs in enumerate(nei):
        out[i, : len(nbs)] = nbs
    _PAD_CACHE[n] = out
    return out


def bridge_table(n: int) -> tuple[tuple[tuple[int, int, int], ...], ...]:
    """For each cell, tuples of (partner, carrier_a, carrier_b) flat indices."""
    cached = _BRIDGE_CACHE.get(n)
    if cached is not None:
        return cached
    table = []
    for r in range(n):
        for c in range(n):
            entries = []
            for (dr, dc), carriers in BRIDGE_OFFSETS:
                pr, pc = r + dr, c + dc
                if not (0 <= pr < n and 0 <= pc < n):
                    continue
                ok = True
                cells = []
                for cr, cc in carriers:
                    ar, ac = r + cr, c + cc
                    if not (0 <= ar < n and 0 <= ac < n):
                        ok = False
                        break
                    cells.append(ar * n + ac)
                if ok:
                    entries.append((pr * n + pc, cells[0], cells[1]))
            table.append(tuple(entries))
    out = tuple(table)
    _BRIDGE_CACHE[n] = out
    return out


def two_distance(board_flat: np.ndarray, n: int, player: int, source: str) -> np.ndarray:
    """Two-distance from one of ``player``'s edges to every cell.

    ``source`` is 'top'/'bottom' for BLACK and 'left'/'right' for WHITE.
    """
    ncells = n * n
    nei = padded_neighbours(n)
    opp = WHITE if player == BLACK else BLACK

    is_own = board_flat == player
    is_opp = board_flat == opp

    rows = np.arange(ncells) // n
    cols = np.arange(ncells) % n
    src = np.full(ncells, INF, dtype=np.float64)
    if source == "top":
        src[rows == 0] = 0.0
    elif source == "bottom":
        src[rows == n - 1] = 0.0
    elif source == "left":
        src[cols == 0] = 0.0
    else:
        src[cols == n - 1] = 0.0
    # An edge cell already owned by `player` sits *on* the source group.
    src[is_opp] = INF

    d = np.full(ncells + 1, INF, dtype=np.float64)
    d[ncells] = INF
    # The source edge occupies *two* of the candidate slots: a board edge
    # cannot be cut, so touching it already counts as a redundant connection.
    # With only one slot the relaxation deadlocks at infinity, because no empty
    # cell can ever find a second finite neighbour to get started from.
    cand = np.empty((ncells, 8), dtype=np.float64)
    for _ in range(2 * n + 4):
        cand[:, :6] = d[nei]
        cand[:, 6] = src
        cand[:, 7] = src
        cand.sort(axis=1)
        new = np.where(is_own, cand[:, 0], cand[:, 1] + 1.0)
        new[is_opp] = INF
        new = np.minimum(new, d[:ncells])
        if np.array_equal(new, d[:ncells]):
            break
        d[:ncells] = new
    return np.minimum(d[:ncells], INF)


def _edges(player: int) -> tuple[str, str]:
    return ("top", "bottom") if player == BLACK else ("left", "right")


def connection_cost(board_flat: np.ndarray, n: int, player: int) -> tuple[float, np.ndarray]:
    """(potential, per-cell cost) for ``player``.

    ``potential`` is the cheapest total two-distance through any single cell --
    a proxy for "how many more stones do I need to connect my two edges".
    """
    e1, e2 = _edges(player)
    d1 = two_distance(board_flat, n, player, e1)
    d2 = two_distance(board_flat, n, player, e2)
    total = d1 + d2
    # A cell is counted by both halves, so an empty one is charged twice.
    total = np.where(board_flat == EMPTY, total - 1.0, total)
    total = np.minimum(total, INF)
    return float(total.min()), total


def evaluate(board: HexBoard, player: int) -> float:
    """Static evaluation in [-1, 1] from ``player``'s point of view."""
    if board.winner:
        return 1.0 if board.winner == player else -1.0
    flat = board.array().reshape(-1)
    opp = WHITE if player == BLACK else BLACK
    my_pot, _ = connection_cost(flat, board.n, player)
    op_pot, _ = connection_cost(flat, board.n, opp)
    if my_pot >= INF:
        return -1.0
    if op_pot >= INF:
        return 1.0
    diff = op_pot - my_pot
    return float(np.tanh(diff / 3.0)) * 0.99  # keep clear of true terminal scores


def move_scores(board: HexBoard, player: int, defence_weight: float = 1.0) -> np.ndarray:
    """Per-cell desirability for ``player`` (higher is better, -inf if illegal).

    A good Hex move usually does two jobs at once: it shortens my own
    connection and lengthens my opponent's.  Scoring both and adding them is
    the classic "attack + defence" move ordering.
    """
    flat = board.array().reshape(-1)
    n = board.n
    opp = WHITE if player == BLACK else BLACK
    _, my_cost = connection_cost(flat, n, player)
    _, op_cost = connection_cost(flat, n, opp)
    score = -np.minimum(my_cost, 1e5) - defence_weight * np.minimum(op_cost, 1e5)
    # Two-distance is flat on an empty board (every cell costs the same n
    # stones to connect through), so break ties towards the centre -- the
    # standard Hex convention, and correct: central cells keep more options.
    score = score + 0.15 * _centrality(n)
    score[flat != EMPTY] = -np.inf
    return score


_CENTRALITY_CACHE: dict[int, np.ndarray] = {}


def _centrality(n: int) -> np.ndarray:
    cached = _CENTRALITY_CACHE.get(n)
    if cached is not None:
        return cached
    idx = np.arange(n * n)
    r, c = idx // n, idx % n
    mid = (n - 1) / 2.0
    # Chebyshev-ish distance from the centre, normalised to [0, 1] then negated.
    dist = np.maximum(np.abs(r - mid), np.abs(c - mid)) / max(mid, 1e-9)
    out = (1.0 - dist).astype(np.float64)
    _CENTRALITY_CACHE[n] = out
    return out


def immediate_wins(board: HexBoard, player: int) -> list[int]:
    """Cells where ``player`` moving now would complete a connection.

    Read straight off the union-find: a cell wins iff it simultaneously touches
    a group joined to one of ``player``'s edges and a group joined to the
    other.  O(6) per empty cell instead of copying the board 121 times.
    """
    if board.winner:
        return []
    n = board.n
    flat = board.board
    if player == BLACK:
        a_root, b_root = board._find(board.TOP), board._find(board.BOTTOM)
    else:
        a_root, b_root = board._find(board.LEFT), board._find(board.RIGHT)
    if a_root == b_root:
        return []
    out = []
    nei = board._nei
    for cell in range(n * n):
        if flat[cell]:
            continue
        r, c = divmod(cell, n)
        if player == BLACK:
            hits_a, hits_b = r == 0, r == n - 1
        else:
            hits_a, hits_b = c == 0, c == n - 1
        for nb in nei[cell]:
            if flat[nb] != player:
                continue
            root = board._find(nb)
            if root == a_root:
                hits_a = True
            elif root == b_root:
                hits_b = True
        if hits_a and hits_b:
            out.append(cell)
    return out


def broken_bridge_response(board: HexBoard, player: int, last_move: int | None) -> int | None:
    """If the opponent just intruded into one of my bridges, plug the other cell.

    This single pattern is most of what separates a naive Hex bot from a
    respectable one: it makes second-order connections effectively unbreakable.
    """
    if last_move is None:
        return None
    n = board.n
    flat = board.array().reshape(-1)
    opp = WHITE if player == BLACK else BLACK
    if flat[last_move] != opp:
        return None
    bridges = bridge_table(n)
    for cell, entries in enumerate(bridges):
        if flat[cell] != player:
            continue
        for partner, ca, cb in entries:
            if flat[partner] != player:
                continue
            if last_move == ca and flat[cb] == EMPTY:
                return cb
            if last_move == cb and flat[ca] == EMPTY:
                return ca
    return None
